### PR TITLE
Validation of global variables and entry points

### PR DESCRIPTION
**Pull request by kvark**
Another step towards  #59
This PR establishes the structure of the validator, and covers global variables and entry points.

There is an annoying bit that it's hard to say if a global is a storage buffer/texture or not, I wasn't able to figure that out yet, but it's not a blocker.

### CODE REVIEW

File: src/proc/validator.rs [hunk 1]
```diff
@@ -28,8 +81,242 @@ impl Validator {
...
+        if ep.early_depth_test.is_some() && stage != crate::ShaderStage::Fragment {
+            return Err(EntryPointError::UnexpectedEarlyDepthTest);
+        }
+        if stage == crate::ShaderStage::Compute {
+            if ep
+                .workgroup_size
+                .iter()
+                .any(|&s| s == 0 || s > MAX_WORKGROUP_SIZE)
+            {
+                return Err(EntryPointError::InvalidWorkgroupSize);
```
Comment (Napokue) [inline src/proc/validator.rs:211]: Shouldn't the error be something like: `OutOfRangeWorkgroupSize`? The word `Invalid` seems a little bit too generic here in my opinion.

File: src/proc/validator.rs [hunk 2]
```diff
@@ -28,8 +81,242 @@ impl Validator {
...
+        for ((var_handle, var), &usage) in module
+            .global_variables
+            .iter()
+            .zip(&ep.function.global_usage)
+        {
+            if usage.is_empty() {
+                continue;
+            }
+
+            println!("var {:?}", var); //TEMP
```
Comment (Napokue) [inline src/proc/validator.rs:229]: This can be removed? :)

File: src/proc/validator.rs [hunk 3]
```diff
@@ -28,8 +81,242 @@ impl Validator {
...
+                    //TODO: built-in checks?
+                    let mut storage_usage = crate::GlobalUse::empty();
+                    if var.storage_access.contains(crate::StorageAccess::LOAD) {
+                        storage_usage |= crate::GlobalUse::LOAD;
+                    }
+                    if var.storage_access.contains(crate::StorageAccess::STORE) {
+                        storage_usage |= crate::GlobalUse::STORE;
+                    }
+                    //TODO: make this more explicit somehow
+                    if storage_usage.is_empty() {
```
Comment (Napokue) [inline src/proc/validator.rs:290]: I don't really understand this statement. What needs to happen when the user uses an Uniform or StorageBuffer storage class? Don't we just need to explicitly state that this is a syntax error?

Or is it possible to not have a storage usage specified beforehand?

**Review by Napokue — CHANGES_REQUESTED** [review]
Thank you! I've got a couple of things.

File: src/proc/validator.rs (hunk 2 above)
Comment (kvark) [reply src/proc/validator.rs:229]: wow, this is why we need reviews :)
thank you for spotting!

File: src/proc/validator.rs (hunk 3 above)
Comment (kvark) [reply src/proc/validator.rs:290]: I'll remove the todo for now. Technically, the storage access flags being non-zero means that it's a storage, so IR is actually fine.

**Pull request merged**